src/v/redpanda/admin_server.cc

```diff
@@ -5778,16 +5779,124 @@ void admin_server::register_wasm_transform_routes() {
+      std::move(name));
+    co_await throw_on_error(*req, ec, model::controller_ntp);
+    co_return ss::json::json_void();
 }
 
 ss::future<ss::json::json_return_type>
 admin_server::list_transforms(std::unique_ptr<ss::http::request>) {
-    throw ss::httpd::bad_request_exception("data transforms not enabled");
+    if (!_transform_service->local_is_initialized()) {
+        throw ss::httpd::bad_request_exception("data transforms not enabled");
+    }
+    throw ss::httpd::server_error_exception("unimplemented");
 }
 
-ss::future<std::unique_ptr<ss::http::reply>> admin_server::deploy_transform(
-  std::unique_ptr<ss::http::request>, std::unique_ptr<ss::http::reply>) {
-    throw ss::httpd::bad_request_exception("data transforms not enabled");
+namespace {
+json::validator make_transform_deploy_validator() {
+    const std::string schema = R"(
+{
+    "type": "object",
+    "properties": {
+        "name": {
+            "type": "string"
+        },
+        "input_topic": {
+            "type": "string"
+        },
+        "output_topics": {
+            "type": "array",
+            "items": {
+              "type": "string"
+            }
+        },
+        "environment": {
+            "type": "array",
+            "items": {
+              "type": "object",
+              "properties": {
+                  "key": {
+                      "type": "string"
+                  },
+                  "value": {
+                      "type": "string"
+                  }
+              },
+              "required": [
+                  "key",
+                  "value"
+              ],
+              "additionalProperties": false
+            }
+        }
+    },
+    "required": ["name", "input_topic", "output_topics"],
+    "additionalProperties": false
+}
+)";
+    return json::validator(schema);
+}
+} // namespace
+
+ss::future<ss::json::json_return_type>
+admin_server::deploy_transform(std::unique_ptr<ss::http::request> req) {
+    if (!_transform_service->local_is_initialized()) {
+        throw ss::httpd::bad_request_exception("data transforms not enabled");
+    }
+    // The request body could be large here, so stream it into an iobuf.
+    iobuf body;
+    auto out_stream = make_iobuf_ref_output_stream(body);
+    co_await ss::copy(*req->content_stream, out_stream);
+    // Now wrap the iobuf in a stream and parse the header JSON document out.
+    iobuf_istreambuf ibuf(body);
+    std::istream stream(&ibuf);
+    json::IStreamWrapper s(stream);
+    // Make sure to stop when the JSON object is over, as after is the Wasm
+    // binary.
+    json::Document doc;
+    doc.ParseStream<
+      rapidjson::kParseDefaultFlags | rapidjson::kParseStopWhenDoneFlag>(s);
+    if (doc.HasParseError()) {
+        throw ss::httpd::bad_request_exception(
+          fmt::format("JSON parse error: {}", doc.GetParseError()));
+    }
+    // Drop the JSON object from the iobuf, so this is just our Wasm binary.
+    body.trim_front(s.Tell());
```

Comment:
interesting. how robust do you think this is? i guess in principle it should be robust. if the binary came first we could prefix it with a length, or maybe even use http multi-part headers or be inspired by that?

Reply:
It is robust for the implementation in RPK: https://github.com/redpanda-data/redpanda/blob/d6b8984cd062d4dfafb7a8e1e563bfb7037bd14b/src/go/rpk/pkg/adminapi/api_transform.go#L32

In general if you're using cURL it should be:
```
echo "{}" > request.bin
cat transform.wasm >> request.bin
curl -d @request.bin ...
```

> if the binary came first we could prefix it with a length

Wasm is self terminating, but we'd need a parser for it in Redpanda (not that hard but still).

> use http multi-part 

TBH if there was an HTTP multi-part parser already available I would just no brainer do that, but this is mostly born out of not wanting to write a parser for the multi-part and this was the simplest thing outside of that.

Do you think we should switch to multipart uploads?

Reply:
nah, i was just wary of the json library being able to give the correct offset to seek to in the iobuf.